python/ray/util/sgd/torch/training_operator.py

```diff
@@ -109,13 +146,197 @@ def _set_timers(self, timers):
+        training components (Cuda, DDP, Distributed Sampler, Fp16).
+        You do not need to handle GPU/devices in this function; ``Ray SGD``
+        will do that for you.
+
+        If more than one model, optimizer, or scheduler is passed in,
+        you should implement your own custom training loop.
+
+        .. code-block:: python
+
+            @override(TrainingOperator)
+            def setup(self, config):
+                model = ...
+                optimizer = ...
+                train_loader = ...
+                val_loader = ...
+                loss = ...
+
+                self.model, self.optimizer, self.train_loader,
+                self.val_loader, self.loss = self.register(models=model,
+                optimizers=optimizer, train_loader=train_loader,
+                validation_loader=validation_loader, criterion=loss)
+
+                # At this point DDP, Cuda, Distributed Sampling, and Fp16
+                # are set up for all our components. We then use self.model,
+                # self.optimizer, etc. in our training loop.
+
+
+        Args:
+            models (torch.nn.Module or Iterable[nn.Module]): Pytorch model or
+                multiple Pytorch models to use for training. If
+                `use_gpu=True` is passed into ``TorchTrainer``, and Cuda is
+                available, models will automatically be placed on GPU.
+                If ``wrap_ddp=True`` is passed into ``TorchTrainer``,
+                models will be wrapped in DDP. If wrap_ddp is False,
+                you should handle DDP for your models in setup.
+            optimizers (torch.optim.Optimizer or Iterable[
+                torch.optim.Optimizer]): Pytorch optimizer or multiple Pytorch
+                optimizers to use for training.
+            train_loader (Iterator): An iterator for training
+                data. If None is explicitly passed in, a Ray SGD Dataset
+                must be passed in through TorchTrainer.train. Ray SGD will
+                automatically use a Distributed Sampler if TorchTrainer(...,
+                add_dist_sampler=True).
+            validation_loader (Iterator): An iterator for validation
+                data. Ray SGD will automatically use a Distributed Sampler
+                if TorchTrainer(..., add_dist_sampler=True).
+            criterion (Callable, optional): Function to return loss
+                metric given features and target. If not provided,
+                must implement a custom training loop.
+            schedulers (torch.optim.lr_scheduler or Iterable[
+                torch.optim.lr_scheduler], optional): A learning rate
+                scheduler or multiple learning rate schedulers.
+
+        Returns:
+            Tuple of model, optimizer, criterion if not None, and scheduler
+            if not None. train_loader and validation_loader are not
+            returned. You should use the iterators passed into train_epoch
+            and validate instead.
```

Comment:
I think the semantics of this function are a little confusing. Instead, you should break this into two calls:

```
self.register(model, optim, sched, criterion)
self.register_data(train_loader, validation_loader)
```
Then, it's explicit that one set of items will passthrough, and one set will not.

Reply:
Agreed. Split it up into register and register_data.